### PR TITLE
Add arborio library

### PR DESCRIPTION
**Pull request by noraabiakar**
Create an `arborio` library that contains methods `load_swc_allen` (already implemented and moved into `arborio`)  and `load_swc_neuron`. These generate `segment_tree`s compliant with the allen institute specs and neuron's specs. 
In cases where the specs are unclear or buggy, we raise exceptions. 
Also add unit tests for both methods. 

Addresses #1189

### CODE REVIEW

**Comment by noraabiakar** [comment]
bors try

**Review by thorstenhater — CHANGES_REQUESTED** [review]

File: arborio/include/arborio/swcio.hpp [hunk 1]
```diff
@@ -0,0 +1,56 @@
+#pragma once
+
+#include <arbor/swcio.hpp>
```
[the remark this answers is not in the document]
Comment (noraabiakar) [reply arborio/include/arborio/swcio.hpp:3]: That's a good point. I wasn't sure about what should/shouldn't go in arborio at first. But now I'm thinking all swc code should go in there.

File: arborio/include/arborio/swcio.hpp [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+struct swc_unsupported_tag: swc_error {
+    explicit swc_unsupported_tag(int record_id);
+};
+
+// No gaps allowed
+struct swc_unsupported_gaps: swc_error {
+    explicit swc_unsupported_gaps(int record_id);
+};
+
+arb::segment_tree load_swc_neuron(const std::vector<arb::swc_record>& records);
```
[the remark this answers is not in the document]
Comment (noraabiakar) [reply arborio/include/arborio/swcio.hpp:53]: I'll add some functions to the docs. And I'll add some comments here explaining briefly what the functions do.

File: arborio/swcio.cpp [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+        if (r.tag == soma_tag) {
+            if (prev_tag != soma_tag)   throw swc_non_consecutive_soma{r.id};
+            if (prev_id != r.parent_id) throw swc_non_serial_soma{r.id};
+            soma_records.push_back(r);
+        }
+
+        // Find record index of the parent
+        auto parent_iter = record_index.find(r.parent_id);
+
+        if (parent_iter == record_index.end() || records[parent_iter->second].id == r.id) throw arb::bad_swc_data{r.id};
```
[the remark this answers is not in the document]
Comment (noraabiakar) [reply arborio/swcio.cpp:82]: Right. I adopted this error message from the `load_swc_allen` function. And was too lazy to replace all its usages. But I will now.

File: arborio/swcio.cpp [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+        {
+            throw arb::bad_swc_data{r.id};
+        }
+
+        // Assert that all samples have the same tag as their parent, except those attached to the soma.
+        if (p != soma_id && r.tag != records[parent_iter->second].tag) {
+            throw swc_mismatched_tags{r.id};
+        }
+
+        // Assert that all non-root samples have a tag of 2, 3, or 4.
```
[the remark this answers is not in the document]
Comment (noraabiakar) [reply arborio/swcio.cpp:256]: This function was originally written by Ben based on his understanding of the allen swc conventions. I think the point is that they only allow single point somas and axons, dendrites or apical dendrites. So 1 would be an error here, as well as all tags > 4

File: arborio/swcio.cpp [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+    arb::segment_tree tree;
+
+    // Model the spherical soma as a cylinder with length=2*radius.
+    // The cylinder is centred on the origin, and extended along the y axis.
+    double soma_rad = sloc.radius;
+    tree.append(arb::mnpos, {0, -soma_rad, 0, soma_rad}, {0, soma_rad, 0, soma_rad}, 1);
+
+    // Build branches off soma.
+    std::unordered_map<int, arb::msize_t> smap; // SWC record id -> segment id
+    std::set<int> unused_samples;
```
[the remark this answers is not in the document]
Comment (noraabiakar) [reply arborio/swcio.cpp:279]: Yes

File: arborio/swcio.cpp [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+    for (const auto& r: records) {
+        int id = r.id;
+        if (id == soma_id) continue;
+
+        // If sample i has the root as its parent don't create a segment.
+        if (r.parent_id == soma_id) {
+            if (no_gaps) {
+                // Assert that this branch starts on the "surface" of the spherical soma.
+                auto d = std::fabs(soma_rad - std::sqrt(r.x * r.x + r.y * r.y + r.z * r.z));
+                if (d > 1e-3) { // 1 nm tolerance
```
[the remark this answers is not in the document]
Comment (noraabiakar) [reply arborio/swcio.cpp:289]: Again, this was written by Ben. We can change it, but I'll ask him first.

File: python/morphology.cpp [hunk 7]
```diff
@@ -8,101 +8,38 @@
...
 #include <arbor/swcio.hpp>
 
+#include <arborio/swcio.hpp>
+
 #include "error.hpp"
 #include "strprintf.hpp"
 
 namespace pyarb {
 
+arb::segment_tree load_swc_neuron(const std::string& fname) {
```
[the remark this answers is not in the document]
Comment (noraabiakar) [reply python/morphology.cpp:18]: Yep, this will go in the docs

File: arborio/swcio.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (thorstenhater) [reply arborio/swcio.cpp:256]: Yes, it should be an error, but of the non-consecutive soma kind.

[the remark this answers is not in the document]
Comment (thorstenhater) [reply arborio/swcio.cpp:256]: I am less sure about the other tags.

File: arborio/swcio.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (thorstenhater) [reply arborio/swcio.cpp:279]: Ok, good, maybe change that name, but I am fine with not doing that.

File: arborio/swcio.cpp [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+    // Vector of records that make up the soma
+    std::vector<arb::swc_record> soma_records = {soma_prox};
+    int prev_tag = soma_prox.tag;
+    int prev_id = soma_prox.id;
+
+    for (std::size_t i = 1; i < records.size(); ++i) {
+        const auto& r = records[i];
+        record_index[r.id] = i;
+
+        if (r.tag == soma_tag) {
```
[the remark this answers is not in the document]
Comment (noraabiakar) [reply arborio/swcio.cpp:73]: I would prefer to keep the preliminary error checking, and building the `record_index` in the same loop

File: arborio/swcio.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (noraabiakar) [reply arborio/swcio.cpp:256]: I'll split this error into: `multi-point soma`, and `unsupported tag` errors.

File: arborio/swcio.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (noraabiakar) [reply arborio/swcio.cpp:279]: I'll add a comment.

File: arborio/swcio.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (noraabiakar) [reply arborio/swcio.cpp:477]: After considering this, I think making the tolerance a parameter would probably cause some problems. Some users may choose 0 as the tolerance which is not a good idea; or they'll want to allow gaps and will have to choose an arbitrarily large tolerance. I think keeping it a `bool` is probably easier. What do you think?

File: arborio/include/arborio/swcio.hpp [hunk 9]
```diff
@@ -38,9 +38,59 @@ struct swc_spherical_soma: swc_error {
     explicit swc_spherical_soma(int record_id);
 };
 
-// Bad or inconsistent SWC data was fed to an `swc_data` consumer.
-struct bad_swc_data: swc_error {
-    explicit bad_swc_data(int record_id);
+// Smells like a non-spherical soma.
+struct swc_non_spheical_soma: swc_error {
```
Comment (thorstenhater) [inline arborio/include/arborio/swcio.hpp:42]: Typo.

File: arborio/swcio.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (thorstenhater) [reply arborio/swcio.cpp:480]: I see your point, but currently the cut-off is quite arbitrarily 1nm. Considering the way way these files are generate, ie some person clicking and dragging samples on an image of prepared tissue, there might be need of a larger or smaller tolerance depending on the source. 

A solution could be to wrap the tolerance in a new-type, like we do with the cable_cell_parameters and signal the user when an unreasonable value is chosen.

That said: noone has asked for it explicitely.

File: doc/concepts/morphology.rst [hunk 10]
```diff
@@ -482,6 +482,96 @@ because it has two children: the dendrites attached to its distal end.
...
+.. code:: Python
+
+   # id, tag, x, y,   z, r, parent
+      1,   1, 0, 0,   0, 5, -1
+      2,   1, 0, 0,  10, 5,  1
+      3,   2, 0, 0, -10, 2,  1
+      4,   3, 0, 0, 100, 1,  2
+
+Samples 1 and 2 will form the soma; samples 1 and 3 will form the axon, connected to the soma at the proximal end;
+samples 2 and 4 will form the dendrite, connected to the soma at the distal end.
```
Comment (thorstenhater) [inline doc/concepts/morphology.rst:539]: I would love to see Ben's sketches here.

File: doc/concepts/morphology.rst [hunk 11]
```diff
@@ -482,6 +482,96 @@ because it has two children: the dendrites attached to its distal end.
...
+In addition to the previously mentioned checks, the Allen interpretation expects a single-sample soma to be the first
+sample of the file and to be interpreted as a spherical soma. Arbor represents the spherical soma as a cylinder with
+length and diameter equal to the diameter of the sample representing the sphere.
+
+This interpretation also expects that samples have the same tag as their parent samples, with the exception of samples
+that have the soma sample as a parent. In this case, when a sample's parent is the soma, no *segment* is created
+between the 2 samples; instead there is a gap in the morphology (represented electrically as a zero-resistance wire).
+Samples with the soma as a parent start new segments, that connect to the distal end of the soma if they are dendrites,
+or to the proximal end of the soma if they are axons or apical dendrites. Only axons, dendrites and apical dendrites
+(tags 2, 3 and 4 respectively) are allowed in this interpretation.
```
Comment (thorstenhater) [inline doc/concepts/morphology.rst:552]: Om the ris of being annoying, perhaps add the soma?

File: arborio/swcio.cpp [hunk 12]
```diff
@@ -0,0 +1,503 @@
...
+    return parse_swc(is, mode);
+}
+
+swc_data parse_swc(std::vector<swc_record> records, swc_mode mode) {
+    swc_data data;
+    data.records = sort_and_validate_swc(std::move(records), mode);
+    return data;
+}
+
+arb::segment_tree as_segment_tree(const std::vector<swc_record>& records) {
```
Comment (thorstenhater) [inline arborio/swcio.cpp:206]: For consistency, maybe this should be called `load_swc_arbor`?

File: doc/concepts/morphology.rst [hunk 13]
```diff
@@ -482,6 +482,96 @@ because it has two children: the dendrites attached to its distal end.
...
+its parent sample; and if more than 1 sample have the same parent sample, the parent sample is interpreted as a fork
+point in the morphology, and acts as the proximal point to a new branch for each of its "child" samples.
+There a couple of exceptions to these rules which are listed below.
+
+Arbor interpretation:
+"""""""""""""""""""""
+In addition to the previously listed checks, the arbor interpretation explicitly disallows SWC files where the soma is
+described by a single sample. It constructs the soma from 2 or more samples that form 1 or more segments in the segment
+tree. A *segment* is constructed between a sample and its parent, and inherits the sample's tag, regardless of
+whether or not its tag matches the parent's tag. This means that there are no gaps in the resulting segment tree.
```
Comment (thorstenhater) [inline doc/concepts/morphology.rst:523]: How is this rule (about segment tags) in the general case?

**Review by thorstenhater — COMMENTED** [review]
Looks good, some small questions left.

File: doc/concepts/morphology.rst (hunk 10 above)
Comment (noraabiakar) [reply doc/concepts/morphology.rst:539]: I considered that but decided it wasn't really necessary. But I can add one if you like.

File: doc/concepts/morphology.rst (hunk 11 above)
Comment (noraabiakar) [reply doc/concepts/morphology.rst:552]: Yes, I'll do that

File: arborio/swcio.cpp (hunk 12 above)
Comment (noraabiakar) [reply arborio/swcio.cpp:206]: I am planning on opening another PR right after this to change this and a few other things. I don't want them to be part of this PR cause it's not gonna show the diff.

Comment (thorstenhater) [reply arborio/swcio.cpp:206]: 👍

File: doc/concepts/morphology.rst (hunk 10 above)
Comment (thorstenhater) [reply doc/concepts/morphology.rst:539]: I think it helps, but maybe just refer/link to the single cell docs?

**Review by thorstenhater — APPROVED** [review]
Looks good.

**Pull request merged**